dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/dataset/metadata/inputRenderer/suggestion/GrantSuggestionDao.java

```diff
@@ -3,10 +3,93 @@
 import javax.ejb.Stateless;
 import javax.persistence.EntityManager;
 import javax.persistence.PersistenceContext;
+import javax.persistence.TypedQuery;
+import java.util.List;
+import java.util.Map;
 
 @Stateless
 public class GrantSuggestionDao {
 
     @PersistenceContext(unitName = "VDCNet-ejbPU")
     private EntityManager em;
+
+    // -------------------- LOGIC --------------------
+
+    /**
+     * Function used to find suggestions according to filters.
+     * @param filteredBy filters that are going to be concatenated with AND.
+     * @param suggestionSourceFieldName target field for query
+     * @param suggestionSourceFieldValue value that is going to be filtered with LIKE %value% and UPPER(value)
+     * @param queryLimit limits the amount of values returned
+     * @return list of suggestions.
+     */
+    public List<String> fetchSuggestions(Map<String, String> filteredBy,
+                                         String suggestionSourceFieldName,
+                                         String suggestionSourceFieldValue,
+                                         int queryLimit) {
+
+        if (filteredBy.isEmpty()){
+            return fetchSuggestions(suggestionSourceFieldName, suggestionSourceFieldValue, queryLimit);
+        }
+
+        String filters = generateFilters(filteredBy);
+
+        TypedQuery<String> query = em.createQuery("SELECT DISTINCT grant." + suggestionSourceFieldName + " FROM GrantSuggestion grant WHERE "
```

Comment:
This will probably break when you pass not empty map that resolves to empty filters. Generated sql will have `WHERE AND {name}` in it which is incorrect. 

Maybe resolve `filters` before `if` and add `|| StringUtils.isEmpty(filters)` to the check?

Reply:
Well shouldn't it break if someone is passing incomplete info?

Reply:
If business-wise this is unexpected, then it's better to throw explicit exception in such case.

Reply:
Well if there is empty value we return empty results in SuggestionInputFieldRenderer.

Reply:
Only if map is empty, but if map resolves to empty filters we will have sql error and IMO some proper exception should be thrown then.

Reply:
I meant SuggestionInputFieldRenderer class where we have `filteredValues.containsValue(StringUtils.EMPTY) ? new HashMap<>()`

Reply:
Ok. But now I wonder whether it's sql-injection safe or not. Theoretically you could pass in filters whole sql statement, but on the other hand we are in control of passed values. However dunno how will it work in API..

Reply:
From all the examples I saw they suggest using .setParameter to be injection safe, so I assumed it should be ok.